src/dbnode/persist/fs/migration/migration.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package migration
+
+import (
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/storage"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+// Task interface is implemented by tasks that wish to perform a data migration
+// on a fileset. This typically involves updating files in a fileset that were created by
+// a previous version of the database client.
+type Task interface {
+	// Run is the set of steps to successfully complete a migration.
+	Run() error
+}
+
+// toVersion1_1Task is an object responsible for migrating a fileset to version 1.1.
+type toVersion1_1Task struct {
+	newMergerFn       fs.NewMergerFn
+	infoFileResult    fs.ReadInfoFileResult
+	shard             uint32
+	namespaceMetadata namespace.Metadata
+	persistManager    persist.Manager
+	sOpts             storage.Options
+	fsOpts            fs.Options
+}
+
+// ShouldMigrateToVersion1_1 indicates whether the fileset should be migrated to 1.1 or not.
+func ShouldMigrateToVersion1_1(info fs.ReadInfoFileResult) bool {
+	return info.Info.MajorVersion == 1 && info.Info.MinorVersion == 0
+}
+
+// NewToVersion1_1Task creates a task for migrating a fileset to version 1.1.
+func NewToVersion1_1Task(opts TaskOptions) (Task, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+	return &toVersion1_1Task{
+		newMergerFn:       opts.NewMergerFn(),
+		infoFileResult:    opts.InfoFileResult(),
+		shard:             opts.Shard(),
+		namespaceMetadata: opts.NamespaceMetadata(),
+		persistManager:    opts.PersistManager(),
+		sOpts:             opts.StorageOptions(),
+		fsOpts:            opts.FilesystemOptions(),
```

Comment:
nit: if we're pulling this all out here anyway, maybe it may be easier to have `toVersion1_1Task` contain just the top level `TaskOptions`, and then `Run()` can extract them itself?

Reply:
Yeah, that's probably true. Makes it less cumbersome to add new options too.